### PR TITLE
[dask] [python] Store co-local data parts as dicts instead of lists

### PR DESCRIPTION
**Pull request by ffineis**
Addresses #3795. The `_train` function in `dask.py` places delayed "parts," i.e. `data`, `label`, `sample_weight`, and `group` data into lists (which get distributed to workers) to enforce co-locality of data.  Each worker is distributed a `list_of_parts`, a list of lists. Once the worker receives its lists of parts, it infers the `data`, `label`, and then optionally `sample_weight` and `group` from each `list_of_parts` by assuming that `data` is in position 0, `label` is in position 1, and if the model is *not* a DaskLGBMRanker, the second position is reserved for `sample_weight`.

This is an issue of readability + scalability - implying the presence of `sample_weight` and `group` based on the length of each sublist of `list_of_parts` in conjunction with the estimator type is... hard to understand and doesn't lend itself well to storing additional `parts` like `eval_sets`. Instead, let's just store individual sets of parts as dicts instead of lists. Now each worker will receive its portion of the overall dataset as a list of dicts instead of a list of lists.

### CODE REVIEW

File: python-package/lightgbm/dask.py [hunk 1]
```diff
@@ -197,29 +196,20 @@ def _train(client, data, label, params, model_factory, sample_weight=None, group
...
     data_parts = _split_to_parts(data=data, is_matrix=True)
     label_parts = _split_to_parts(data=label, is_matrix=False)
+    parts = [{'data': x, 'label': y} for (x, y) in zip(data_parts, label_parts)]
 
     if sample_weight is not None:
         weight_parts = _split_to_parts(data=sample_weight, is_matrix=False)
-    else:
-        weight_parts = None
+        for i, d in enumerate(parts):
+            parts[i] = {**d, 'weight': weight_parts[i]}
```
Comment (jameslamb) [inline python-package/lightgbm/dask.py:207]: ~Will this result in each of the existing parts (data, label) in `d` being copied when they're passed back in? I think it might.~ Ok ignore that, I tested and was wrong about how `**` worked.

Still, I have a slight preference just modifying `parts[i]` directory, taking advantage of the fact that we know it's already a dictionary. 

I think this is a little easier to read, what do you think?

```suggestion
        for i in len(parts):
            parts[i]['weight'] = weight_parts[i]
```

**Review by jameslamb — CHANGES_REQUESTED** [review]
This is great, thank you! I just have one very small nitpicky style thing.

File: python-package/lightgbm/dask.py (hunk 1 above)
Comment (StrikerRUS) [reply python-package/lightgbm/dask.py:207]: I absolutely agree with James! But it should be `for i in range(len(parts)):`

Comment (jameslamb) [reply python-package/lightgbm/dask.py:207]: ha! typed too fast, thanks. Edited it.

File: python-package/lightgbm/dask.py [hunk 2]
```diff
@@ -197,29 +196,20 @@ def _train(client, data, label, params, model_factory, sample_weight=None, group
...
-    if weight_parts is None and group_parts is None:
-        parts = zip(data_parts, label_parts)
-    elif weight_parts is not None and group_parts is None:
-        parts = zip(data_parts, label_parts, weight_parts)
-    elif weight_parts is None and group_parts is not None:
-        parts = zip(data_parts, label_parts, group_parts)
-    else:
-        parts = zip(data_parts, label_parts, weight_parts, group_parts)
+        for i, d in enumerate(parts):
+            parts[i] = {**d, 'group': group_parts[i]}
```
Comment (jameslamb) [inline python-package/lightgbm/dask.py:212]:
```suggestion
        for i in range(len(parts)):
            parts[i]['group'] = group_parts[i]
```

File: python-package/lightgbm/dask.py [hunk 3]
```diff
@@ -136,25 +136,24 @@ def _train_part(params, model_factory, list_of_parts, worker_address_to_port, re
...
     try:
         model = model_factory(**params)
-
         if is_ranker:
-            group = _concat(parts[-1])
-            if len(parts) == 4:
-                weight = _concat(parts[2])
-            else:
-                weight = None
             model.fit(data, y=label, sample_weight=weight, group=group, **kwargs)
```
Comment (StrikerRUS) [inline python-package/lightgbm/dask.py:155]: I believe this whole `try` context can be simplified to single call of `model.fit(data, y=label, sample_weight=weight, group=group, **kwargs)`.
`group` param can be passed to the base `LGBMModel` class without the need to check whether it is ranker or not (`is_ranker`).

File: python-package/lightgbm/dask.py (hunk 1 above)
Comment (ffineis) [reply python-package/lightgbm/dask.py:207]: yep yeah, this is easier to understand, all for it

File: python-package/lightgbm/dask.py (hunk 3 above)
Comment (ffineis) [reply python-package/lightgbm/dask.py:155]: Yep this is true, good catch. This was a holdover from the world before `fit` was estimator-type-agnostic and each Dask model subclass had its own fit method. Will change!

Comment (ffineis) [reply python-package/lightgbm/dask.py:155]: Wait no - sorry I take prior comment back. This is a call to the *local*/serial model `fit` method. Classifier/Regressor don't accept `type` argument: ```Exception: TypeError("fit() got an unexpected keyword argument 'group'")```. So I don't think that `group` can be safely passed as a kwarg to all model types.

Comment (StrikerRUS) [reply python-package/lightgbm/dask.py:155]: Oh sorry, indeed! I forgot about that `LGBMClassifier` and `LGBMRegressor` overwrite `fit` method and `group` is not available there.

File: python-package/lightgbm/dask.py (hunk 2 above)
Comment (ffineis) [reply python-package/lightgbm/dask.py]: Addressed in 0d36cc4

File: python-package/lightgbm/dask.py (hunk 1 above)
Comment (ffineis) [reply python-package/lightgbm/dask.py]: Addressed in 0d36cc4

**Review by jameslamb — APPROVED** [review]
This is much cleaner and I think it will help reduce the risk of mistakes in the future. Thanks so much!

**Review by StrikerRUS — APPROVED** [review]
Awesome! Thanks! Code is much cleaner now.
